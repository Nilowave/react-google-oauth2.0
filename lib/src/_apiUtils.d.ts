/** @internal */
export interface IPayload {
    readonly email: string;
    readonly code: string;
    readonly scope: string;
}
/** @internal */
export interface IApiResponseData {
    readonly access_token: string;
}
/** @internal */
export declare function postToExchangeApiUrl(apiUrl: string, payload: IPayload): Promise<IApiResponseData>;
/** @internal */
export declare function serverResponse(props: any): void;
/** @internal */
export declare function storeAccessToken(token: string): void;
/**
 * @example
 * ```
 *  if(isLoggedIn()) { // returns true is accessToken exists in LocalStorage
 *      // user logged code...
 *  }
 * ```
 */
export declare function isLoggedIn(): boolean;
/**
 * @example
 * ```
 *  logOutOAuthUser() // removes the accessToken from LocalStorage
 * ```
 * @return void
 */
export declare function logOutOAuthUser(): void;
export declare function _getAccessToken(): string | null;
/**
 * @example
 * ```
 *  fetch(url, {
 *      headers: createOAuthHeaders(),
 *  })
 * ```
 * @return Object
 */
export declare function createOAuthHeaders(): Object;
