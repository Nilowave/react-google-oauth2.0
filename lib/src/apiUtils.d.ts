/** @internal */
export declare function serverResponse(props: any): void;
