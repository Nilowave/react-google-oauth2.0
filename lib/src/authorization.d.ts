/** @internal */
export declare enum GoogleAPIConnectionStrings {
    GOOGLE_OAUTH_ENDPOINT = "https://accounts.google.com/o/oauth2/v2/auth"
}
/** @public */
interface IAuthorizationBase {
    clientId: string;
    /** Where you React app should redirect to if user has successfully logged in **/
    redirectUri: string;
    responseType?: string;
    /** Valid parameter values are online, which is the default value,
     * and offline  Set the value to offline if your application needs
     * to refresh access tokens when the user is not present at the browser
     */
    accessType?: string;
    state?: string;
    includeGrantedScopes?: boolean;
    loginHint?: string;
    prompt?: string;
}
/** @public */
export interface IAuthorizationOptions extends IAuthorizationBase {
    scopes: Array<string>;
}
/** @internal */
export interface IAuthorization {
    params: IAuthorizationOptions;
    /** Set to  `["openid", "profile"]` by default **/
    scopesStr: string;
}
/** @internal */
export declare class Authorization implements IAuthorization {
    private _googleRedirectURL?;
    queryParamsCode?: string;
    params: IAuthorizationOptions;
    scopesStr: string;
    constructor(params: IAuthorizationOptions, scopesStr: string);
    /**
     * @internal
     * @property googleRedirectURL Access the url for the initial
     * browser call to Google's OAuth 2.0 server. Useful for debugging.
     */
    get googleRedirectURL(): string | undefined;
    set googleRedirectURL(value: string | undefined);
    /** @internal */
    createAuthorizationRequestURL(): void;
    /** @internal */
    static createScopes(scopes: Array<string>): string;
    /** @internal */
    redirect: (e: any) => void;
}
export {};
