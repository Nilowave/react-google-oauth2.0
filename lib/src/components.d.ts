import * as React from "react";
import { IAuthorizationOptions } from "../src";
export interface IGoogleButton extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    /** Placeholder image displayed next to button text */
    readonly placeholder?: string;
    /** Remove default styles. The placeholder prop has no effect if placeholder is set to false */
    readonly defaultStyle?: boolean;
    /** See IAuthorizationRequestParams */
    readonly options: IAuthorizationOptions;
    /** Could be your preloader or any other dumb component */
    readonly callback?: () => React.ReactHTMLElement<any>;
    /** The url of the api to perform the exchange */
    readonly apiUrl: string;
}
/** @internal */
interface IGoogleAuthContext {
    readonly queryParamsCode: boolean;
}
export declare const GoogleAuth: React.Provider<IGoogleAuthContext>;
export declare const GoogleAuthConsumer: React.Consumer<IGoogleAuthContext>;
/** @internal */
export declare const InnerButton: (props: IGoogleButton & {
    error?: string;
}) => JSX.Element;
/**
 * @example
 * **Quick Start:**
 *
 * First create an options object that implements an {@link  IAuthorizationOptions} type.
 * Check the {@link  IAuthorizationOptions} and {@link  IAuthorizationBase} types for
 * all required properties. Then, pass the options to the {@link GoogleButton} component.
 *
 * ```IAuthorizationOptions
 *  const options:  = {
 *      clientId: (process.env.CLIENT_ID as string),
 *       redirectUri: "http://localhost:3000",
 *       scopes: ["openid", "profile", "email"],
 *       includeGrantedScopes: true,
 *       accessType: "offline",
 *   };
 *
 *   <GoogleButton
 *         placeholder="demo/search.png"
 *         options={options}
 *         apiUrl="http://localhost:5000/google_login"
 *   />
 * ```
 * @param props see IGoogleButton
 * @constructor
 */
export declare const GoogleButton: (props: IGoogleButton) => JSX.Element | null;
export {};
