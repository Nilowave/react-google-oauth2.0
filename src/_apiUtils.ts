/** @internal */
export interface IPayload {
    readonly email: string;
    readonly code: string;
    readonly scope: string;
}
/** @internal */
export interface IApiResponseData { readonly access_token: string; }
/** @internal */
export async function postToExchangeApiUrl(apiUrl: string, payload: IPayload): Promise<IApiResponseData> {
    const res: Response = await fetch(apiUrl, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
    });
    return res.json();
}
/** @internal */
export function serverResponse(props: any): void {
    const { email = "", code, apiUrl, scope } = props;
    const payload: IPayload = { code, email, scope };
    postToExchangeApiUrl(apiUrl, payload)
        .then((data: IApiResponseData) => {
            // update responseState accessToken
            if(!("access_token" in data)) {
                props.setResponseState({
                    error: "No access_token in response data!",
                });
            } else {
                props.setResponseState({
                    accessToken: data.access_token,
                });
            }

        })
        .catch(err => {
            props.setResponseState({
                error: err.message,
            })
        });
}
/** @internal */
export function storeAccessToken(token: string): void {
     window.localStorage.setItem("accessToken", token);
}

/**
 * @example
 * ```
 *  if(isLoggedIn()) { // returns true is accessToken exists in LocalStorage
 *      // user logged code...
 *  }
 * ```
 */
export function isLoggedIn(): boolean {
    return !!window.localStorage.getItem("accessToken");
}

/**
 * @example
 * ```
 *  logOutOAuthUser() // removes the accessToken from LocalStorage
 * ```
 * @return void
 */
export function logOutOAuthUser(): void {
    window.localStorage.removeItem("accessToken");
}

export function _getAccessToken() {
    return window.localStorage.getItem("accessToken");
}

/**
 * @example
 * ```
 *  fetch(url, {
 *      headers: createOAuthHeaders(),
 *  })
 * ```
 * @return Objects
 */
export function createOAuthHeaders(): Object {
    return {
        "Content-type": "application/jsons",
        "X-Auth-Token": `Bearer ${_getAccessToken()}`,
    }
}

/** @internal */
export function removeOAuthQueryParams(): void {
    const currentLocation = document.location.href;
    if(document.location.search) {
        const clean_uri = currentLocation.substring(0, currentLocation.indexOf("?"));
        window.history.replaceState({}, document.title, clean_uri);
    }
}
