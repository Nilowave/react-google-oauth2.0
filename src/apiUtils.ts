/** @internal */
import {Dispatch, SetStateAction} from "react";

export interface IPayload {
    readonly code: string;
    readonly scope: string;
}
/** @internal */
export interface IApiResponseData { readonly access_token: string; }
/** @internal */
export async function postToExchangeApiUrl(apiUrl: string, payload: IPayload): Promise<IApiResponseData> {
    const res: Response = await fetch(apiUrl, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
    });
    return res.json();
}
/** @internal */
export interface IServerResponseState { readonly accessToken?: string; error?: string}
/** @internal */
export interface IServerResponseProps {
    readonly code: string;
    readonly scope: string;
    readonly apiUrl: string;
    setResponseState: Dispatch<SetStateAction<IServerResponseState>>;
}
/** @internal */
export function serverResponse(props: IServerResponseProps): void {
    const { code, apiUrl, scope } = props;
    const payload: IPayload = { code, scope };
    postToExchangeApiUrl(apiUrl, payload)
        .then((data: IApiResponseData) => {
            // update responseState accessToken
            if(!("access_token" in data)) {
                props.setResponseState({
                    error: "so access_token in response data!",
                });
            } else {
                props.setResponseState({
                    accessToken: data.access_token,
                });
            }

        })
        .catch(err => {
            props.setResponseState({
                error: err.message,
            })
        });
}
/** @internal */
export function storeAccessToken(token: string): void {
     window.localStorage.setItem("accessToken", token);
}

/**
 * @example
 * ```
 *  if(isLoggedIn()) { // returns true is accessToken exists in LocalStorage
 *      // user logged code...
 *  }
 * ```
 */
export function isLoggedIn(): boolean {
    return !!window.localStorage.getItem("accessToken");
}

/**
 * @example
 * ```
 *  logOutOAuthUser() // removes the accessToken from LocalStorage
 * ```
 * @return void
 */
export function logOutOAuthUser(): void {
    window.localStorage.removeItem("accessToken");
}

/**
 * @description Get the stored accessToken
 * @return The Access Token or none
 */
export function getAccessToken(): string | null {
    return window.localStorage.getItem("accessToken");
}

/**
 * @description Warning: Please make sure that as of version `0.0.23` we have removed
 * the `application/json` content type headers from this function. This may cause an issue
 * if you originally didn't construct your request content type headers with `application/json`.
 * @example
 * ```
 *  fetch(url, {
 *      headers: createOAuthHeaders(),
 *  })
 * ```
 *
 * If you require your server to handle authenticating multiple users across many resources
 * (or tables) then pass in the name of the resource, for Example:
 * @example
 * ```
 *  fetch(url, {
 *      headers: createOAuthHeaders("users"),
 *  })
 * ```
 *
 *  With the resource value, the following headers are constructed:
 *
 *  @example
 *  ```
 *  {
 *    "X-Auth-Token" : "<TOKEN>",
 *    "X-Auth-Resource": "users",
 *  }
 *  ```
 * @param resource Optional resource name to look up on the server
 * @return Objects
 */
export function createOAuthHeaders(resource?: string): Object {
    let resourceHeaders = {};
    if(resource) {
        resourceHeaders = {
            "X-Auth-Resource": resource,
        }
    }
    return {
        ...resourceHeaders,
        "X-Auth-Token": `Bearer ${getAccessToken()}`,
    }
}

/** @internal */
export function removeOAuthQueryParams(): void {
    const currentLocation = document.location.href;
    if(document.location.search) {
        const clean_uri = currentLocation.substring(0, currentLocation.indexOf("?"));
        window.history.replaceState({}, document.title, clean_uri);
    }
}
