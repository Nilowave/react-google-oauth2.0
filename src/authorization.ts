/** @internal */
export enum GoogleAPIConnectionStrings {
    GOOGLE_OAUTH_ENDPOINT = "https://accounts.google.com/o/oauth2/v2/auth",
}
/** @public */
interface IAuthorizationBase {
    /** This is your client ID */
    clientId: string;
    /** Where you React app should redirect to if user has successfully logged in **/
    redirectUri: string;
    /** By default this is set to *code* */
    responseType?: string;
    /**
     * Indicates if the api **apiUrl** {@link IGoogleButton} can refresh tokens on the user's
     * behalf. The default value is *online* but if your backend api will refresh the token then
     * you must pass in a string value of *offline* (This is recommended for react apps connected
     * to a REST api backend).
     */
    accessType?: string;
    /**
     * This option is currently unavailable but will be in a added in a future version,
     * TODO - See https://github.com/joegasewicz/react-google-oauth2.0/issues/3
     * */
    state?: string;
    /**
     * Defaulted to true. Any extra scopes the logged in user previously granted
     * will be available within the new access token.
     * */
    includeGrantedScopes?: boolean;
    /** Not required, optional. For prefilling email fields etc. */
    loginHint?: string;
    /**
     * Not required, optional. A space-delimited, case-sensitive list of prompts to
     * present the user. If you don't specify this parameter, the user will be prompted
     * only the first time your project requests access.
     * */
    prompt?: string;
}
/** @public */
export interface IAuthorizationOptions extends IAuthorizationBase {
    scopes: Array<string>;
}
/** @internal */
export interface IAuthorization {
    params: IAuthorizationOptions;
    /** Set to  `["openid", "profile"]` by default **/
    scopesStr: string;
}
/** @internal */
export class Authorization implements IAuthorization {

    private _googleRedirectURL?: string;

    public params: IAuthorizationOptions;

    public scopesStr: string;

    constructor(params: IAuthorizationOptions, scopesStr: string) {
        this.params = params;
        this.scopesStr = scopesStr;
    }

    /**
     * @internal
     * @property googleRedirectURL Access the url for the initial
     * browser call to Google's OAuth 2.0 server. Useful for debugging.
     */
    public get googleRedirectURL() {
        return this._googleRedirectURL;
    }

    public set googleRedirectURL(value: string | undefined) {
        this._googleRedirectURL = value;
    }

    /** @internal */
    public createAuthorizationRequestURL(): void {
        const {
            accessType,
            includeGrantedScopes = true,
            responseType = "code",
            state = null,
            redirectUri,
            clientId,
        } = this.params;
        const domain = GoogleAPIConnectionStrings.GOOGLE_OAUTH_ENDPOINT;
        let url: string;
        url = `${domain}?scope=${this.scopesStr}&`;
        url = accessType ? `${url}access_type=${accessType}&` : url;
        url = `${url}include_granted_scopes=${includeGrantedScopes}&`;
        url = `${url}response_type=${responseType}&`;
        url = state ? `${url}state=${state}&` : url;
        url = `${url}redirect_uri=${redirectUri}&`;
        url = `${url}client_id=${clientId}`;
        this.googleRedirectURL = url;
    }

    /** @internal */
    public static createScopes(scopes: Array<string>): string {
        let str = "";
        scopes.map((scope, i) => {
            if (scopes.length - 1 === i) {
                str += `${scope}`;
            } else {
                str += `${scope}%20`;
            }
        });
        return str;
    }

    /** @internal */
    public redirect = (e: any): void => {
        if(this.googleRedirectURL) {
            window.location.replace(this.googleRedirectURL)
        } else {
            throw new ReferenceError("Error creating redirect url to Google's authorization server");
        }
    }
}