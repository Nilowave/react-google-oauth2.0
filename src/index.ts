import {
    Authorization,
    GoogleAPIConnectionStrings,
    IAuthorizationOptions,
} from "./authorization";

import {
    GoogleButton,
    GoogleAuth,
    GoogleAuthConsumer,
    IGoogleButton,
} from "./components";

export  {
    Authorization,
    GoogleAPIConnectionStrings,
    GoogleButton,
    GoogleAuth,
    GoogleAuthConsumer,
    IGoogleButton,
    IAuthorizationOptions,
}
