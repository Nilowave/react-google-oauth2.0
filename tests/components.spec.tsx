import React from "react";
import renderer from "react-test-renderer";
import {GoogleButton, IAuthorizationOptions} from "../src";
const googleComponents = require("../src/apiUtils");
import {act} from 'react-test-renderer';


describe("#GoogleButton", () => {
    delete window.location;
    (window.location as any) = jest.mock;
    (window.location.replace as any) = (x: string) => window.location.href = x;

    const testParamsOne: IAuthorizationOptions = {
        clientId: "<CLIENT_ID>",
        redirectUri: "http://localhost:3000",
        scopes: ["openid", "profile", "email"],
        includeGrantedScopes: true,
    };

    it("should render a component", () => {
        const component =  <GoogleButton
                  placeholder="demo/search.png" // Optional
                  options={testParamsOne}
                  apiUrl="http://localhost:5000/google_login"
                  defaultStyle={true} // Optional
              />
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should contain text", () => {
        const component =  <GoogleButton
            placeholder="demo/search.png" // Optional
            options={testParamsOne}
            apiUrl="http://localhost:5000/google_login"
            defaultStyle={true} // Optional
        >Sign in with google</GoogleButton>

        const googleButton = renderer.create(component);
        const button = googleButton.root;
        expect(button.findByType("button").props.children).toEqual("Sign in with google");
    });

    it("should contain styles", () => {
        const component =  <GoogleButton
            placeholder="demo/search.png" // Optional
            options={testParamsOne}
            apiUrl="http://localhost:5000/google_login"
            defaultStyle={true} // Optional
        />

        const googleButton = renderer.create(component);
        const button = googleButton.root;
        expect(button.findByType("button").props.style.backgroundSize).toEqual("20px 20px");
        expect(button.findByType("button").props.style.backgroundRepeat).toEqual("no-repeat");
        expect(button.findByType("button").props.style.backgroundPosition).toEqual("12px 10px");
        expect(button.findByType("button").props.style.textIndent).toEqual("19px");
        expect(button.findByType("button").props.style.border).toEqual("1px solid #bdc3c7");
        expect(button.findByType("button").props.style.padding).toEqual("9px 23px");
        expect(button.findByType("button").props.style.borderRadius).toEqual("9px");
        expect(button.findByType("button").props.style.backgroundColor).toEqual("#bdc3c7");
        expect(button.findByType("button").props.style.fontSize).toEqual("18px");
        expect(button.findByType("button").props.style.backgroundImage).toEqual("url(demo/search.png)");
    });

    it("should contain option props", () => {
        const component =  <GoogleButton
            placeholder="demo/search.png" // Optional
            options={testParamsOne}
            apiUrl="http://localhost:5000/google_login"
            defaultStyle={true} // Optional
        />

        const googleButton = renderer.create(component);
        expect(googleButton.root.props.placeholder).toEqual("demo/search.png");
        expect(googleButton.root.props.apiUrl).toEqual("http://localhost:5000/google_login");
        expect(googleButton.root.props.defaultStyle).toEqual(true);
        expect(googleButton.root.props.options.clientId).toEqual("<CLIENT_ID>");
        expect(googleButton.root.props.options.redirectUri).toEqual("http://localhost:3000");
        expect(googleButton.root.props.options.scopes).toEqual(["openid", "profile", "email"]);
        expect(googleButton.root.props.options.includeGrantedScopes).toEqual(true);
    });

    it("should redirect to Googles login page", () => {
        const component =  <GoogleButton
            placeholder="demo/search.png" // Optional
            options={testParamsOne}
            apiUrl="http://localhost:5000/google_login"
            defaultStyle={true} // Optional
        />

        const googleButton = renderer.create(component);
        const button = googleButton.root.findByType("button");
        button.props.onClick();
        const expected = "https://accounts.google.com/o/oauth2/v2/auth?" +
            "scope=openid%20profile%20email&" +
            "access_type=offline&" +
            "include_granted_scopes=true&" +
            "response_type=code&" +
            "redirect_uri=http://localhost:3000&" +
            "client_id=<CLIENT_ID>";
        expect(window.location.href).toEqual(expected);
    });

    it("should send a request to the backend API and return loading component", async () => {
        window.location.search = "http://localhost:3000?email=<EMAIL>&code=<CODE>&scope=<SCOPES>";
        let googleButton;
        const spy = jest.fn();
        await act(async () => {

            const component =  <GoogleButton
                placeholder="demo/search.png" // Optional
                options={testParamsOne}
                apiUrl="http://localhost:5000/google_login"
                defaultStyle={true} // Optional
            />
            googleButton  = renderer.create(component);

            jest.spyOn(googleComponents,  "serverResponse")
                .mockImplementation(spy);

        });
        expect.assertions(2);
        expect((googleButton as any).root.children[0]).toEqual("Loading...");
        // TODO update this test with at least 1 call to serverResponse
        expect(googleComponents.serverResponse).toHaveBeenCalledTimes(0);

    });

    xit("should set token on response from API", async () => {
        // TODO Need updating to test the useContext updates
        const spy = jest.fn();
        await act( async () => {

        const component =  <GoogleButton
            placeholder="demo/search.png" // Optional
            options={testParamsOne}
            apiUrl="http://localhost:5000/google_login"
            defaultStyle={true} // Optional
        />
        const googleButton = renderer.create(component);

        jest.spyOn(React,  "useContext")
            .mockImplementation(() => [{isAuthenticated: true}, jest.fn()]);
        jest.spyOn(googleComponents, "storeAccessToken")
            .mockImplementation(spy);
        });

        expect(spy).toBeCalledTimes(1);
        expect(spy).toBeCalledWith("<access_token>");
    });

    xit("should not recall exchange server if access token is in localStorage", () => {
         // TODO Need updating to test the useContext updates
        const spy1 = jest.fn();
        const spy2 = jest.fn();
        jest.spyOn(googleComponents, "isLoggedIn")
            .mockImplementation(() => true);
        jest.spyOn(React,  "useState")
                .mockImplementation(() => [{accessToken: "<access_token>"}, jest.fn()]);
        jest.spyOn(googleComponents, "serverResponse")
            .mockImplementation(spy1);

        jest.spyOn(googleComponents, "storeAccessToken")
            .mockImplementation(spy2);

        const component =  <GoogleButton
            placeholder="demo/search.pnz" // Optional
            options={testParamsOne}
            apiUrl="http://localhost:5000/google_login"
            defaultStyle={true} // Optional
        />

        const googleButton = renderer.create(component);
        expect(spy1).toBeCalledTimes(0);
        expect(spy2).toBeCalledTimes(0);
    });

});