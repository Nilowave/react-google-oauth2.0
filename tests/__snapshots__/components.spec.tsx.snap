// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#GoogleButton should render a component 1`] = `
<button
  apiUrl="http://localhost:5000/google_login"
  className="google-oauth-btn"
  defaultStyle={true}
  onClick={[Function]}
  options={
    Object {
      "clientId": "<CLIENT_ID>",
      "includeGrantedScopes": true,
      "redirectUri": "http://localhost:3000",
      "scopes": Array [
        "openid",
        "profile",
        "email",
      ],
    }
  }
  placeholder="demo/search.png"
  style={
    Object {
      "backgroundColor": "#bdc3c7",
      "backgroundImage": "url(demo/search.png)",
      "backgroundPosition": "12px 10px",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "20px 20px",
      "border": "1px solid #bdc3c7",
      "borderRadius": "9px",
      "fontSize": "18px",
      "padding": "9px 23px",
      "textIndent": "19px",
    }
  }
/>
`;
